lib/upload.js

```diff
@@ -1014,6 +1097,96 @@ function splitSizeIntoParts (totalSize, partCount) {
   return parts
 }
 
+/**
+ * Build the chunk staggers.
+ *
+ * NOTE: Must be kept in sync with `splitSizeIntoParts` in skynet-js.
+ */
+function buildChunkStaggers (totalSize, partCount, chunkSize) {
+  // The lists of chunks for each part.
+  const partsChunks = []
+  for (let i = 0; i < partCount; i++) {
+    partsChunks.push([])
+  }
+  // The leftover size that must go into the last part.
+  const hasLeftover = (totalSize % chunkSize) > 0
+
+  function newChunk (index) {
+    const partChunks = partsChunks[index]
+
+    let chunkResolve
+    const promise = new Promise((resolve) => chunkResolve = resolve)
+    const chunk = {
+      promise,
+      resolve         : chunkResolve,
+      resolved        : false,
+      nextChunkStagger: null,
+      nextChunkInPart : null,
+      indexInPart     : partChunks.length,
+    }
+
+    // Previous chunk in part should point to this chunk.
+    if (partChunks.length > 0) {
+      partChunks[partChunks.length - 1].nextChunkInPart = chunk
+    }
+
+    // Push chunk to part.
+    partChunks.push(chunk)
+  }
+
+  // Assign chunks to parts in order, looping back to the beginning if we get to
+  // the end of the parts array.
+  let lastPart = -1
+  for (let i = 0; i < Math.floor(totalSize / chunkSize); i++) {
+    newChunk(i % partCount)
+
+    if (i > lastPart) lastPart = i
+  }
+
+  if (hasLeftover) {
+    let lastIndex
+    if (lastPart === -1) {
+      // No parts were visited, so assign to the last part.
+      lastIndex = partCount - 1
+    } else {
+      // Assign the leftover to the part after the last part that was visited, or
+      // the last part in the array if all parts were used.
+      lastIndex = Math.min(lastPart + 1, partCount - 1)
+    }
+    newChunk(lastIndex)
+  }
+
+  // Now assign the stagger order. We start at the part with the most chunks, so
+  // we don't end up with any final pauses.
+  let longestPartIndex = null
+  let longestPartLength = 0
+  for (let i = 0; i < partsChunks.length; i++) {
+    if (longestPartIndex === null || partsChunks[i].length > longestPartLength) {
+      longestPartIndex = i
+      longestPartLength = partsChunks[i].length
+    }
+  }
+  // Get the chunk at which to start uploading.
+  const firstChunk = partsChunks[longestPartIndex][0]
+  // Build the stagger sequence.
+  let lastChunk = null
+  for (let i = 0; i < longestPartLength; i++) {
+    for (let j = longestPartIndex; j < longestPartIndex + partCount; j++) {
+      const index = j % partCount
+      if (partsChunks[index].length > i) {
```

Comment:
this is the one that's confusing me, don't you want to start `i` at the `longestPartLength` and go down? I think the algorithm is pretty neat but it's just missing more information, unit tests would've helped as well as they often also provide some visual insight in what input should yield what output

Reply:
Added some comments and will add unit tests, good catch.

Reply:
Well this project doesn't have unit tests or any infra setup for running them. Gross. 😬

I tried adding the setup myself but it ended up being a lot of changes, including having to rename the babel config file.

https://stackoverflow.com/questions/59879689/jest-syntaxerror-cannot-use-import-statement-outside-a-module

I agree with @kwypchlo that we should keep changes to a minimum, unless we plan to just totally break from upstream (which maybe we do? The code is very poor; there are functions calling themselves recursively to upload more - and other horrors.)

Anyways, I didn't add the unit tests, but the existing integration tests do verify that the chunks are staggered in the correct order.

Reply:
works for me, didn't realize you were working in a fork to be honest ^^